core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala

```diff
@@ -0,0 +1,746 @@
+        )
+      })
+
+      _ <- MigrateAction.raiseWhen(billingProject.invalidBillingAccount) {
+        WorkspaceMigrationException(
+          message = "The migration failed: invalid billing account on billing project",
+          data = errorDetails ++ Map(
+            ("billingProject" -> billingProject.projectName),
+            ("billingProjectBillingAccount" -> billingProjectBillingAccount)
+          )
+        )
+      }
+
+      _ <- MigrateAction.raiseWhen(workspaceBillingAccount != billingProjectBillingAccount) {
+        WorkspaceMigrationException(
+          message = "Workspace migration failed: billing account on workspace differs from billing account on billing project",
+          data = errorDetails ++ Map(
+            ("workspaceBillingAccount" -> workspaceBillingAccount),
+            ("billingProject" -> billingProject.projectName),
+            ("billingProjectBillingAccount" -> billingProjectBillingAccount)
+          )
+        )
+      }
+
+      workspaceService <- MigrateAction.asks(_.workspaceService)
+      (googleProjectId, googleProjectNumber) <- MigrateAction.liftIO {
+        workspaceService.setupGoogleProject(
+          billingProject,
+          workspaceBillingAccount,
+          workspace.workspaceId,
+          workspace.toWorkspaceName
+        ).io
+      }
+
+      configured <- nowTimestamp
+      _ <- inTransaction { _ =>
+        workspaceMigrations
+          .filter(_.id === migration.id)
+          .map(r => (r.newGoogleProjectId, r.newGoogleProjectNumber, r.newGoogleProjectConfigured))
+          .update((googleProjectId.value.some, googleProjectNumber.value.some, configured.some))
+      }
+    } yield ()
+
+
+  final def createTempBucket: MigrateAction[Unit] =
+    for {
+      migration <- findMigration { m =>
+        m.newGoogleProjectConfigured.isDefined && m.tmpBucketCreated.isEmpty
+      }
+
+      googleProjectId <- MigrateAction.liftIO(IO {
+        migration.newGoogleProjectId.getOrElse(throw noGoogleProjectError(migration))
+      })
+
+      workspace <- getWorkspace(migration.workspaceId)
+      tmpBucketName <- MigrateAction.liftIO(randomSuffix("terra-workspace-migration-"))
+      _ <- createBucketInSameRegion(
+        migration,
+        GcsBucketName(workspace.bucketName),
+        googleProjectId,
+        GcsBucketName(tmpBucketName)
+      )
+
+      created <- nowTimestamp
+      _ <- inTransaction { _ =>
+        workspaceMigrations
+          .filter(_.id === migration.id)
+          .map(r => (r.tmpBucket, r.tmpBucketCreated))
+          .update((tmpBucketName.some, created.some))
+      }
+    } yield ()
+
+
+  final def issueWorkspaceBucketTransferJob: MigrateAction[Unit] =
+    for {
+      migration <- findMigration { m =>
+        m.tmpBucketCreated.isDefined && m.workspaceBucketTransferJobIssued.isEmpty
+      }
+
+      tmpBucketName <- MigrateAction.liftIO(IO {
+        migration.tmpBucketName.getOrElse(throw noGoogleProjectError(migration))
+      })
```

Comment:
Here's where my knowledge of Scala breaks down and my head starts to hurt because we're mixing imperitive and functional code. It turns out that in a for-comp it's the same: Expanding the operations:
```scala
ReaderT.flatMap(fa, f)
    = Reader { env => OptionT.flatMap(fa(env), value => f(value)(env).value) }
    = Reader { env => OptionT {  IO.flatMap(fa(env).value, option => if (option.isEmpty) IO.pure(None) else ((value => f(value)(env).value) option.get) }
```
Then `IO.flatMap` does a try-catch. So yes. In this case, it's the same!
            


.flatMap {